tests/integration/test_forward_auth.py

```diff
@@ -0,0 +1,188 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+import json
+import logging
+from os.path import join
+
+import pytest
+import requests
+import yaml
+from lightkube import Client
+from lightkube.resources.core_v1 import ConfigMap
+from pytest_operator.plugin import OpsTest
+from tenacity import retry, stop_after_attempt, wait_exponential
+
+from tests.integration.conftest import deploy_traefik_if_not_deployed
+
+logger = logging.getLogger(__name__)
+
+OATHKEEPER_CHARM = "oathkeeper"
+TRAEFIK_CHARM = "traefik-k8s"
+IAP_REQUIRER_CHARM = "iap-requirer"
+
+
+@pytest.fixture(scope="module")
+def lightkube_client(ops_test: OpsTest) -> Client:
+    client = Client(field_manager=OATHKEEPER_CHARM, namespace=ops_test.model.name)
+    return client
+
+
+async def get_app_address(ops_test: OpsTest, app_name: str) -> str:
+    """Get address of an app."""
+    status = await ops_test.model.get_status()  # noqa: F821
+    return status["applications"][app_name]["public-address"]
+
+
+async def get_reverse_proxy_app_url(
+    ops_test: OpsTest, ingress_app_name: str, app_name: str
+) -> str:
+    """Get the ingress address of an app."""
+    address = await get_app_address(ops_test, ingress_app_name)
+    proxy_app_url = f"http://{address}/{ops_test.model.name}-{app_name}/"
+    logger.debug(f"Retrieved address: {proxy_app_url}")
+    return proxy_app_url
+
+
+@pytest.mark.skip_if_deployed
+@pytest.mark.abort_on_fail
+async def test_deployment(ops_test: OpsTest, traefik_charm, forward_auth_tester_charm):
+    """Deploy the charms and integrations required to set up an Identity and Access Proxy."""
+    await deploy_traefik_if_not_deployed(ops_test, traefik_charm)
+
+    # Enable experimental-forward-auth
+    await ops_test.model.applications[TRAEFIK_CHARM].set_config(
+        {"enable_experimental_forward_auth": "True"}
+    )
+
+    # Deploy oathkeeper
+    await ops_test.model.deploy(
+        OATHKEEPER_CHARM,
+        channel="latest/edge",
+        config={"dev": "True"},
+        trust=True,
+    )
+
+    # Deploy the iap-requirer charm with integrations
+    await ops_test.model.deploy(
+        application_name=IAP_REQUIRER_CHARM,
+        entity_url=forward_auth_tester_charm,
+        resources={"oci-image": "kennethreitz/httpbin"},
+        trust=True,
+    )
+
+    await ops_test.model.integrate(f"{IAP_REQUIRER_CHARM}:ingress", TRAEFIK_CHARM)
+    await ops_test.model.integrate(f"{IAP_REQUIRER_CHARM}:auth-proxy", OATHKEEPER_CHARM)
+
+    await ops_test.model.integrate(f"{TRAEFIK_CHARM}:experimental-forward-auth", OATHKEEPER_CHARM)
+
+    await ops_test.model.wait_for_idle(
+        [TRAEFIK_CHARM, OATHKEEPER_CHARM, IAP_REQUIRER_CHARM], status="active", timeout=1000
+    )
+
+
+@retry(
```

Comment:
Could you please add a comment why the retry is needed after the `wait_for_idle` above?
Can we use `block_until` some condition flips, e.g. oathkeeper endpoint?